pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,464 @@
+
+    def _single_shot_expval_gradients(self, expval_cost, args, kwargs):
+        """Compute the single shot gradients of an ExpvalCost object"""
+
+        qnodes = expval_cost.qnodes
+        coeffs = expval_cost.hamiltonian.coeffs
+        device = qnodes[0].device
+
+        self.check_device(device)
+        original_shots = device.shots
+
+        if self.lipschitz is None:
+            self.check_learning_rate(coeffs)
+
+        try:
+            if self.wrs:
+                grads = self.weighted_random_sampling(
+                    qnodes, coeffs, self.max_shots, self.trainable_args, *args, **kwargs
+                )
+            else:
+                device.shots = [(1, self.max_shots)]
+                # We iterate over each trainable argument, rather than using
+                # qml.jacobian(expval_cost), to take into account the edge case where
+                # different arguments have different shapes and cannot be stacked.
+                grads = [
+                    qml.jacobian(expval_cost, argnum=i)(*args, **kwargs)
+                    for i in self.trainable_args
+                ]
+        finally:
+            device.shots = original_shots
+
+        return grads
+
+    def _single_shot_qnode_gradients(self, qnode, args, kwargs):
+        """Compute the single shot gradients of a QNode."""
+        device = qnode.device
+
+        self.check_device(qnode.device)
+        original_shots = device.shots
+
+        if self.lipschitz is None:
+            self.check_learning_rate(1)
+
+        try:
+            device.shots = [(1, self.max_shots)]
+            grads = [qml.jacobian(qnode, argnum=i)(*args, **kwargs) for i in self.trainable_args]
+        finally:
+            device.shots = original_shots
+
+        return grads
+
+    def compute_grad(
+        self, objective_fn, args, kwargs
+    ):  # pylint: disable=signature-differs,arguments-differ
+        r"""Compute gradient of the objective function, as well as the variance of the gradient,
+        at the given point.
+
+        Args:
+            objective_fn (function): the objective function for optimization
+            args: arguments to the objective function
+            kwargs: keyword arguments to the objective function
+
+        Returns:
+            tuple[array[float], array[float]]: a tuple of NumPy arrays containing the gradient
+            :math:`\nabla f(x^{(t)})` and the variance of the gradient.
+        """
+        if isinstance(objective_fn, qml.ExpvalCost):
+            grads = self._single_shot_expval_gradients(objective_fn, args, kwargs)
+        elif isinstance(objective_fn, qml.QNode) or hasattr(objective_fn, "device"):
+            grads = self._single_shot_qnode_gradients(objective_fn, args, kwargs)
+        else:
+            raise ValueError(
+                "The objective function must either be encoded as a single QNode or "
+                "an ExpvalCost object for the Rosalin optimizer. "

```

Comment:
Aren't we calling this the "Shot adaptive optimizer" rather than "Rosalin"?
```suggestion
                "an ExpvalCost object for the Shot adaptive optimizer. "
```

Reply:
good catch!